zip-0222.rst

```diff
@@ -0,0 +1,322 @@
+form ``tze_data`` described below. Serialization of all integral and vector
+types is as with Bitcoin.
+
+``tze_data`` encoding:
+
+====================== ==================================== =====================
+Field                  Description                          Type           
+====================== ==================================== =====================
+``tze_id``             extension ``type``                   ``compactSize``
+``tze_mode``           extension ``mode``                   ``compactSize``
+``tze_data_len``       length of precondition/witness data  ``compactSize``
+``tze_data``           serialized precondition/witness data ``witness_len`` bytes
+====================== ==================================== =====================
+
+TZE Input Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``prevout_hash``       hash of previous txid                ``uint256``    
+``prevout_in``         index into previous txn's outputs    ``uint32``
+``witness``            witness for prevout's precondition   ``tze_data``
+====================== ==================================== ===============
+
+
+TZE Output Encoding:
+
+====================== ==================================== ===============
+Field                  Description                          Type           
+====================== ==================================== ===============
+``amount``             spendable amount, in zatoshi         ``int64``    
+``precondition``       precondition encumbering ``amount``  ``tze_data``
+====================== ==================================== ===============
+
+
+Consensus rules
+---------------
+
+Once this ZIP becomes active, the following new consensus rules are enforced:
+
+- For each ``(outpoint, witness)`` pair in ``tze_inputs``:
+
+  - ``outpoint`` MUST reference a precondition of the same type and mode in an already-mined
+    transaction.
+  - ``tze_verify(mode, precondition, witness, context)`` MUST return ``true``.
+
+- If a transaction has non-empty ``tze_inputs`` and non-empty ``tze_outputs``, then every
+  element in both fields MUST have the same ``type``.
+
+- Non-coinbase transactions MUST have at least one of the following:
+  - nonempty transparent inputs
+  - nonempty shielded inputs
+  - nonempty ``tze_inputs``
+
+- Transactions MUST have at least one of the following:
+  - nonempty transparent outputs
+  - nonempty shielded outputs
+  - nonempty ``tze_outputs``
+
+- All ``amount`` field values of ``tze_output`` records MUST be nonnegative and not greater
+  than ``MAX_MONEY``.
+
+- The sum of amounts associated with all transparent ouputs and TZE outputs within 
+  a transaction MUST NOT exceed the sum of the amounts associated with transaction inputs
+  plus the sapling ``valueBalance`` amount.
```

Comment:
@daira check the sign of these amounts?

Reply:
This rule is not quite right; compare with my suggested replacement.

There's also an ambiguity of definitions about whether TZE inputs/outputs should be considered as "transparent" inputs/outputs in general. Existing consensus rules were only considering Bitcoin-style transparent inputs and outputs, but TZE inputs/outputs are also "transparent" from a security point of view. We need to consider both user-facing and specification terms carefully here. I would favour adding another adjective in front of "transparent" when we mean non-TZE.

```suggestion
- The sum of amounts going out of the transparent value pool of a transaction (that is, Bitcoin-style outputs and TZE outputs, plus JoinSplit ``vpub_old`` values) MUST NOT exceed the sum of amounts going into that pool (that is, Bitcoin-style inputs and TZE inputs, plus JoinSplit ``vpub_new`` values, plus the Sapling ``valueBalance`` amount).
```

Reply:
I can't seem to apply this suggestion, though I could do so manually. Could you move it to the current HEAD?

Reply:
Never mind, fixed.